src/AWS.Deploy.Common/IO/DirectoryManager.cs

```diff
@@ -1,23 +1,94 @@
+        string[] GetDirectories(string path, string? searchPattern = null, SearchOption searchOption = SearchOption.TopDirectoryOnly);
+        bool IsEmpty(string path);
+        bool ExistsInsideDirectory(string parentDirectoryPath, string childPath);
+        void Delete(string path, bool recursive = false);
     }
 
     public class DirectoryManager : IDirectoryManager
     {
-        public DirectoryInfo CreateDirectory(string path) => Directory.CreateDirectory(path);
+        public DirectoryInfo CreateDirectory(string path)
+        {
+            try
+            {
+                return Directory.CreateDirectory(path);
+            }
+            catch (Exception ex)
+            {
+                throw new FailedToCreateDirectoryException($"Failed to create a directory at path - {path}", ex);
+            } 
+        }
+
+        public DirectoryInfo GetDirectoryInfo(string path)
+        {
+            try
+            {
+                return new DirectoryInfo(path);
+            }
+            catch (Exception ex)
+            {
+                throw new FailedToFindDirectoryInfo($"Failed to find directory info for the path - {path}", ex);
+            }
+        }
 
         public bool Exists(string path) => Directory.Exists(path);
 
-        public string[] GetFiles(string path, string? searchPattern = null) => Directory.GetFiles(path, searchPattern ?? "*");
+        public string[] GetFiles(string path, string? searchPattern = null, SearchOption searchOption = SearchOption.TopDirectoryOnly)
+        {
+            try
+            {
+                return Directory.GetFiles(path, searchPattern ?? "*", searchOption);
+            }
+            catch (Exception ex)
+            {
+                throw new FailedToRetrieveFilesException($"Failed to retrieve files inside the specified directory path - {path}", ex);
+            }
+        }
+
+        public string[] GetDirectories(string path, string? searchPattern = null, SearchOption searchOption = SearchOption.TopDirectoryOnly)
+        {
+            try
+            {
+                return Directory.GetDirectories(path, searchPattern ?? "*", searchOption);
+            }
+            catch (Exception ex)
+            {
+                throw new FailedToRetrieveDirectoriesException($"Failed to retrieve directories inside the specified directory path - {path}", ex);
+            }
+        }
+
+        public bool IsEmpty(string path) => GetFiles(path).Length == 0 && GetDirectories(path).Length == 0;
+
+        public bool ExistsInsideDirectory(string parentDirectoryPath, string childPath)
+        {
+            var parentDirectoryFullPath = GetDirectoryInfo(parentDirectoryPath).FullName;
+            var childFullPath = GetDirectoryInfo(childPath).FullName;
+            return childFullPath.Contains(parentDirectoryFullPath + Path.DirectorySeparatorChar, StringComparison.InvariantCulture);
```

Comment:
Can you add some unit tests for this?

Reply:
Refer to https://github.com/aws/aws-dotnet-deploy/pull/257#discussion_r671526419